### PR TITLE
[DSD]  Fix set_optimizer_state_dict() changes the parameters with some optimizers

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125339
* #125338
* __->__ #125708

Summary:
Some optimizers, like AdamW, change the parameters even if gradients are zero. So `set_optimizer_state_dict()` may affect the parameters values with these optimizers. This PR fixes the issue.

This PR also fixes https://github.com/pytorch/pytorch/issues/121186.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @LucasLLC

### CODE REVIEW

**Review by wz337 — APPROVED** [review]
LGTM!

**Comment by Craigacp** [comment]
Is there going to be a PyTorch 2.3.2, and if so would it be possible to get this fix in it? I've spent all day running down slight parameter differences in my model when loading checkpoints as this is called in `get_optimizer_state_dict` which is necessary to get the optimizer dictionary to load into with dist_cp.load. When I loaded in the optimizer checkpoint it changed the freshly loaded model checkpoint because it stepped an empty optimizer so weight decayed all my parameters.
